### PR TITLE
6913 lock version conflict mitigation

### PR DESCRIPTION
**Pull request by ttoomey**
## Description

[GH Issue ](https://github.com/open-path/Green-River/issues/6913)

Two changes to try and reduce occurrence of users encountering lock version errors. Spot checking suggests this is often due to opening the same record in multiple tabs.

1. add cross-tab sync mechanism. When a tab receives a new version (detected via lock-version), surface that on all other tabs (for clients, enrollments, and assessments) and prompt the user to reload
   * This does not address conflicts due to server side changes
   * users can close snack bar alert
   * known issue: if the user has a form open when the reload, their current form data would be lost. Seems like this could be addressed with improved props handling or maybe some way to temporarily save and re-apply form-state
   * The current implementation uses `BroadcastChannel` which is a newer feature. Looks like it's now supported on evergreen browsers
2. when clicking the button to edit the client record, re-fetch the client data (could extend to other pages)
   * this might catch server side changes
   * we could take this approach on other buttons but the client form is the low hanging fruit

misc fix:
* bug fix for ConfirmationDialog not showing cancelText

##testing 
1. testing cross tab sync: open a two tabs with the same client. Update the client record in one, the other tab should display prompt to reload the client record. Reloading should load new data and close the prompt. Clicking continue should close the prompt but not reload.
2. test client refresh: go to the client dash. In the console or else where, update the client (client.touch in rails). Open client details form and save, there should be no lock-version error

## Type of change
[//]: # 'remove options that are not relevant'
- [x] Bug fix
- [x] New feature (adds functionality)

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (eslint)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: src/components/elements/ConfirmationDialog.tsx [hunk 1]
```diff
@@ -73,7 +73,7 @@ const ConfirmationDialog = ({
               variant='gray'
               data-testid='cancelDialogAction'
             >
-              {cancelText || unconfirmable ? 'Close' : 'Cancel'}
+              {cancelText || (unconfirmable ? 'Close' : 'Cancel')}
```
Comment (ttoomey) [inline src/components/elements/ConfirmationDialog.tsx:76]: looks like a bug where cancelText was not displayed

File: src/modules/enrollment/components/pages/EnrollmentDashboard.tsx [hunk 2]
```diff
@@ -112,6 +113,16 @@ const EnrollmentDashboard: React.FC = () => {
...
+        <LocalVersionCoordinationPrompt
+          recordType='Enrollment'
+          recordId={enrollment.id}
+          currentVersion={enrollment.lockVersion}
+        />
+        <LocalVersionCoordinationPrompt
+          recordType='Client'
+          recordId={client.id}
+          currentVersion={client.lockVersion}
+        />
```
Comment (ttoomey) [inline src/modules/enrollment/components/pages/EnrollmentDashboard.tsx:125]: Could end up with overlapping modals here. We could refactor the LocalVersionCoordinationPrompt to handle that

File: src/modules/localVersionCoordination/hooks/useLocalVersionCoordination.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      }
+    },
+    [currentVersion]
+  );
+
+  // manage channel connection
+  const channelRef = useRef<BroadcastChannel | null>(null);
+  useEffect(() => {
+    const channelName = `record-sync-${recordType}-${recordId}`;
+    channelRef.current = new BroadcastChannel(channelName);
```
Comment (ttoomey) [inline src/modules/localVersionCoordination/hooks/useLocalVersionCoordination.tsx:36]: [BroadcastChannel docs here](https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API)

File: src/modules/assessments/components/IndividualAssessmentPage.tsx [hunk 4]
```diff
@@ -42,15 +44,23 @@ const IndividualAssessmentPage = () => {
...
-      enrollment={enrollment}
-      client={client}
-      viewingDefinition={assessment.definition}
-      editingDefinition={
-        assessment.upgradedDefinitionForEditing || assessment.definition
-      }
-      assessment={assessment}
-    />
+    <>
+      <LocalVersionCoordinationPrompt
```
Comment (ttoomey) [inline src/modules/assessments/components/IndividualAssessmentPage.tsx:47]: if this approach seems good, we'd also need to add to the household enrollments

File: src/modules/localVersionCoordination/LocalVersionCoordinationPrompt.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    }
+  }, [currentVersion, latestVersion]);
+
+  const handleContinue = () => {
+    setAcknowledgedVersion(latestVersion);
+  };
+
+  // hide prompt if we are on the latest version OR we are on an old version but the user has acknowledged it
+  const showPrompt =
+    latestVersion > Math.max(acknowledgedVersion, currentVersion);
```
Comment (ttoomey) [inline src/modules/localVersionCoordination/LocalVersionCoordinationPrompt.tsx:53]: perhaps we should show a snackbar or alert even if the user has acknowledge the stale record

File: src/components/elements/SnackbarAlert.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import { ReactNode } from 'react';
+
+interface Props {
+  open: boolean;
+  onClose: VoidFunction;
+  children?: ReactNode;
+  alertProps?: AlertProps;
+  title?: string;
+}
+const SnackbarAlert: React.FC<Props> = ({
```
Comment (ttoomey) [inline src/components/elements/SnackbarAlert.tsx:12]: Ideally we'd use this in the existing Apollo Alert but refactoring that felt like a bit too much scope creep for this branch

File: src/modules/enrollment/components/pages/EnrollmentDashboard.tsx (hunk 2 above)
Comment (ttoomey) [reply src/modules/enrollment/components/pages/EnrollmentDashboard.tsx:125]: this doesn't seem to be an issue in practice

**Review by gigxz — APPROVED** [review]
This looks great @ttoomey!

**Comment by gigxz** [comment]
Screenshot 2024-12-13 at 12 46 09 PM

https://github.com/greenriver/hmis-frontend/pull/986/commits/5808a545c03411c14ad957783aaeeb18e2b339cc

Linking PR https://github.com/greenriver/hmis-frontend/pull/991 and added a TODO for it, because I will update Alerts to use the grayscale button variant when merged, per design library!
